src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,253 @@
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf16_length <= 134217728, Error,
+              "Expected UTF-16le length is too large to transcode");
+  auto destbuf = kj::heapArray<UChar>(expected_utf16_length);
+
+  size_t actual_length =
+      simdutf::convert_utf8_to_utf16le(source.asChars().begin(), source.size(), destbuf.begin());
+  KJ_ASSERT(actual_length == expected_utf16_length, "Expected UTF16 length mismatch");
+
+  // simdutf returns 0 for invalid UTF-8 value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.asBytes().attach(kj::mv(destbuf));
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUtf8FromUcs2(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  KJ_ASSERT(source.size() % 2 == 0, "UTF-16le input size should be multiple of 2");
+  auto utf16_input =
+      kj::arrayPtr<char16_t>(reinterpret_cast<char16_t*>(source.begin()), source.size() / 2);
+  size_t expected_utf8_length =
+      simdutf::utf8_length_from_utf16le(utf16_input.begin(), utf16_input.size());
+
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf8_length <= 134217728, Error,
+              "Expected UTF-8 length is too large to transcode");
+
+  auto destbuf = kj::heapArray<kj::byte>(expected_utf8_length);
+
+  size_t actual_length = simdutf::convert_utf16le_to_utf8(utf16_input.begin(), utf16_input.size(),
+                                                          destbuf.asChars().begin());
+  KJ_ASSERT(actual_length == expected_utf8_length, "Expected UTF8 length mismatch");
+
+  // simdutf returns 0 for invalid UTF-8 value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.asBytes().attach(kj::mv(destbuf));
+}
+
+} // namespace
+
+Converter::Converter(Encoding encoding, kj::StringPtr substitude) {
+  UErrorCode status = U_ZERO_ERROR;
+  auto name = getEncodingName(encoding);
+  auto conv = ucnv_open(name, &status);
+  KJ_ASSERT(U_SUCCESS(status));
+  conv_ = kj::Own<UConverter>(conv, ConverterDisposer::INSTANCE);
+  setSubstitudeChars(substitude);
+}
+
+Converter::Converter(UConverter* converter, kj::StringPtr substitude)
+    : conv_(converter, ConverterDisposer::INSTANCE) {
+  setSubstitudeChars(substitude);
+}
+
+UConverter* Converter::conv() const {
+  return const_cast<UConverter*>(conv_.get());
+}
+
+size_t Converter::maxCharSize() const {
+  KJ_ASSERT_NONNULL(conv_.get());
+  return ucnv_getMaxCharSize(conv_.get());
+}
+
+size_t Converter::minCharSize() const {
+  KJ_ASSERT_NONNULL(conv_.get());
+  return ucnv_getMinCharSize(conv_.get());
+}
+
+void Converter::reset() {
+  KJ_ASSERT_NONNULL(conv_.get());
+  ucnv_reset(conv_.get());
+}
+
+void Converter::setSubstitudeChars(kj::StringPtr sub) {
+  KJ_ASSERT_NONNULL(conv_.get());
```

Comment:
Per the ICU API, setting the subst chars to zero-length input will fail. I think with Node.js we always use `?` as the default if I am remembering correctly, so maybe the default should be `?` here too instead of zero-length? Otherwise, check for zero-length here and skip the call to setSubstChars if there's no input.

Reply:
Actually, I think it's any input `< ucnv_getMinCharSize(conv_.get())` that will fail